pkg/requester/selection/any.go

```diff
@@ -0,0 +1,129 @@
+		err = requester.NewErrNotEnoughNodes(minCount, len(rankedNodes))
+		return nil, err
+	}
+
+	sort.Slice(rankedNodes, func(i, j int) bool {
+		return rankedNodes[i].Rank > rankedNodes[j].Rank
+	})
+
+	selectedNodes := rankedNodes[:system.Min(len(rankedNodes), desiredCount)]
+	selectedInfos := generic.Map(selectedNodes, func(nr requester.NodeRank) model.NodeInfo { return nr.NodeInfo })
+	return selectedInfos, nil
+}
+
+func (s *anyNodeSelector) SelectNodes(ctx context.Context, job *model.Job) ([]model.NodeInfo, error) {
+	minCount := system.Max(job.Spec.Deal.MinBids, job.Spec.Deal.Concurrency)
+	desiredCount := minCount * int(s.overAskForBidsFactor)
+	return s.Select(ctx, job, minCount, desiredCount)
+}
+
+func (s *anyNodeSelector) SelectNodesForRetry(ctx context.Context, job *model.Job, jobState *model.JobState) ([]model.NodeInfo, error) {
+	// calculate how many executions we still need, and evaluate if we can ask more nodes to bid
+	var minExecutions int
+	if job.Spec.Deal.MinBids > 0 && jobState.ReceivedBidCount() < job.Spec.Deal.MinBids {
+		// if we are still queuing bids, then we need at least MinBids to start accepting bids
+		minExecutions = system.Max(job.Spec.Deal.GetConcurrency(), job.Spec.Deal.MinBids)
+	} else if jobState.PublishedOrPublishingCount() > 0 {
+		// if at least a single execution was published or still publishing, then we don't need to retry in case some executions failed to publish
+		minExecutions = 1
+	} else {
+		// by default, we need executions as many as the job concurrency
+		minExecutions = job.Spec.Deal.GetConcurrency()
+	}
+
+	enoughExecutions := jobState.NonDiscardedCount() >= minExecutions
+	if enoughExecutions {
+		// There are enough running executions, so no action to take right now.
+		return []model.NodeInfo{}, nil
+	}
+
+	desiredNodeCount := minExecutions - jobState.NonDiscardedCount()
+	return s.Select(ctx, job, desiredNodeCount, desiredNodeCount)
+}
+
+// We have over-asked for bids, so now only select the number of nodes we
+// actually need.
+func (*anyNodeSelector) SelectBids(ctx context.Context, job *model.Job, jobState *model.JobState) (accept, reject []model.ExecutionState) {
+	if jobState.ReceivedBidCount() < job.Spec.Deal.MinBids {
+		return nil, nil
+	}
+
+	executionsWaiting := jobState.GroupExecutionsByState()[model.ExecutionStateAskForBidAccepted]
+	requiredNewExecutions := system.Max(0, job.Spec.Deal.GetConcurrency()-jobState.ActiveCount())
+	return executionsWaiting[:requiredNewExecutions], executionsWaiting[requiredNewExecutions:]
+}
+
+// We can mark a job as partially complete if we have a single complete
+// execution, and fully complete once we are over the user's required confidence.
+func (*anyNodeSelector) CanCompleteJob(ctx context.Context, job *model.Job, jobState *model.JobState) (bool, model.JobStateType) {
+	completedCount := jobState.CompletedCount()
+	if completedCount < 1 {
+		return false, jobState.State
+	} else if completedCount < job.Spec.Deal.GetConfidence() {
```

Comment:
Say we have a confidence of 4, 2 successful executions, and 2 failed executions (`ExecutionStateFailed`). I am not sure it's correct to say the execution was partially completed since `JobStateCompletedPartially` means some nodes have failed to publish their results when that may not be the case here - executions can fail for other reason. Do you think there is an opportunity to be more granular here with the returned `JobStateType`?

This comment also applies to the implementation in `allNodeSelector`.

I may also be way off in the weeds here as it could be the case that this method is never called in the circumstances I have described above.

Reply:
I think the decision to be made here is what to do if some/any nodes fail to execute (after retries). Do we:

1. Mark the whole job as `Failed`? After all, the user said run it on all nodes and we couldn't. This is what we do today when the user says "run it on X nodes" and we only manage <X.
2. Mark the whole job as "partially completed" (using some state? could update `CompletedPartially` to be broader than just "results failed to publish" or introduce something new)

Reply:
I'd prefer option one as it keeps behavior inline with the existing expectations.